### PR TITLE
 Fix leave and ret doesn't translate to pseudocode

### PR DESCRIPTION
**Pull request by mrpaja**
It translates to pseudocode now. However, the problem has comprised more than leave and ret. It touched every op code that consists of a single word like leave, return or nop.  I am not sure if you like the pseudo code.

### CODE REVIEW

**Comment by Maijin** [comment]
https://github.com/radare/radare2/issues/8883 ?

**Comment by Maijin** [comment]
When I say `return 0xXXXXX` See how this is done in other "pseudo" code engine:

```
mov        eax, 0x0     ; CODE XREF=sub_401310+120
leave
ret
```

becomes

```
return 0x0;
```

File: libr/parse/p/parse_x86_pseudo.c [hunk 1]
```diff
@@ -172,37 +173,60 @@ static int parse(RParse *p, const char *data, char *str) {
 		ptr = strchr (buf, ' ');
 		if (!ptr)
 			ptr = strchr (buf, '\t');
+		if (!ptr) 
+			ptr = strchr (buf, '\0');
+		if (!ptr)
```
Comment (radare) [inline libr/parse/p/parse_x86_pseudo.c]: open brace here

File: libr/parse/p/parse_x86_pseudo.c [hunk 2]
```diff
@@ -172,37 +173,60 @@ static int parse(RParse *p, const char *data, char *str) {
 		ptr = strchr (buf, ' ');
 		if (!ptr)
 			ptr = strchr (buf, '\t');
+		if (!ptr)
```
Comment (radare) [inline libr/parse/p/parse_x86_pseudo.c]: add missing braces {}

File: libr/parse/p/parse_x86_pseudo.c [hunk 3]
```diff
@@ -172,37 +173,60 @@ static int parse(RParse *p, const char *data, char *str) {
...
+
+		optr = ptr;
+		ptr = strchr (ptr, ',');
 		if (ptr) {
 			*ptr = '\0';
 			for (++ptr; *ptr == ' '; ptr++);
-			strncpy (w0, buf, sizeof (w0) - 1);
-			strncpy (w1, ptr, sizeof (w1) - 1);
-
+			strncpy (w1, optr, sizeof (w1) - 1);
```
Comment (radare) [inline libr/parse/p/parse_x86_pseudo.c]: maybe use r_str_ncpy to ensure end termination?

File: libr/parse/p/parse_x86_pseudo.c [hunk 4]
```diff
@@ -172,37 +173,60 @@ static int parse(RParse *p, const char *data, char *str) {
...
+	
+		const char *wa[] = { w0, w1, w2, w3 };
+		int nw = 0;
+		for (i = 0; i < 4; i++) {
+			if (wa[i][0] != '\0') {
+				nw++;
+			}
+		}
+
+		if (strstr (w1, "ax"))
```
Comment (radare) [inline libr/parse/p/parse_x86_pseudo.c]: open brace {

File: libr/parse/p/parse_x86_pseudo.c [hunk 5]
```diff
@@ -172,37 +173,60 @@ static int parse(RParse *p, const char *data, char *str) {
...
-					*ptr = '\0';
-					for (++ptr; *ptr == ' '; ptr++);
-					strncpy (w2, optr, sizeof (w2) - 1);
-					strncpy (w3, ptr, sizeof (w3) - 1);
-				}
+				strncpy (w2, optr, sizeof (w2) - 1);
+				strncpy (w3, ptr, sizeof (w3) - 1);
 			}
 		}
+
```
Comment (radare) [inline libr/parse/p/parse_x86_pseudo.c]: empty line

File: libr/parse/p/parse_x86_pseudo.c [hunk 6]
```diff
@@ -172,37 +173,60 @@ static int parse(RParse *p, const char *data, char *str) {
...
-			const char *wa[] = { w0, w1, w2, w3 };
-			int nw = 0;
-			for (i = 0; i < 4; i++) {
-				if (wa[i][0] != '\0') {
-					nw++;
-				}
+			if (!(p->retleave_asm = (char *) malloc (32))) {
+				return false;
 			}
+			sprintf (p->retleave_asm, "return %s", w2);
```
Comment (radare) [inline libr/parse/p/parse_x86_pseudo.c]: overflow

File: libr/parse/p/parse_x86_pseudo.c [hunk 7]
```diff
@@ -172,37 +173,60 @@ static int parse(RParse *p, const char *data, char *str) {
...
+			replace (nw, wa, str);
+		} else if (strstr (w0, "leave") && p->retleave_asm) {
+			replace (nw, wa, str);
+		} else if (strstr (w0, "ret") && p->retleave_asm) {
+			strncpy (str, p->retleave_asm, strlen (p->retleave_asm) + 1);
```
Comment (radare) [inline libr/parse/p/parse_x86_pseudo.c]: wat. overflow

File: libr/parse/p/parse_x86_pseudo.c [hunk 8]
```diff
@@ -172,37 +173,60 @@ static int parse(RParse *p, const char *data, char *str) {
...
+			free (p->retleave_asm);
+			p->retleave_asm = NULL;
+		} else {
+			free (p->retleave_asm);
```
Comment (radare) [inline libr/parse/p/parse_x86_pseudo.c]: use R_FREE

**Comment by Maijin** [comment]
Still have:

```
               leave                    //str.Password_OK_:__n
               return 0
```

instead of

```
return 0
```

:___

**Comment by Maijin** [comment]
@mrpaja *

**Comment by mrpaja** [comment]
b800000000     eax = 0
c9                       leave 
c3                       return 0
should be transformed into ? 
b800000000     eax = 0
c9                        
c3                       return 0

@Maijin

**Comment by Maijin** [comment]
would be cleaner yes

**Pull request merged**